### ISSUE TITLE
Feature: HTTP3 loadbalancer for HTTP service - with WebTransport support

### ISSUE THREAD

**Issue by Jashepp** [traefik/traefik#11204]
### Welcome!

- [x] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [x] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

Hi, can we get support for WebTransport within the **HTTP service**, where we can provide a HTTP loadBalancer that connects to a HTTP 3 (Quic) server, which accepts WebTransport (or some other similar solution to get this working)?

Currently the only method of setting up WebTransport with Traefik is by using a **UDP service**, with a UDP loadBalancer, which cannot be done along-side an existing HTTP3 setup on the **same** endpoint, on the **same** TLS host/domain (for example, in #9858 they ended up using a UDP service).

I have tried using `https://address:443/` as a HTTP loadBalancer, but that seems to only connect using TCP HTTP, no UDP HTTP 3.
I can't get it working, even when using an endpoint with http3 enabled, which works for all other requests, and even when udp http3 is available at `address:443`.
So I am assuming that the http loadBalancer does not yet support connecting over UDP HTTP 3.
The service I am trying to have behind Traefik is something [very similar to this](https://socket.io/get-started/webtransport), using nodejs & socketio.
I cannot find any other case of webtransport working through Traefik without using a UDP service.

There is a `scheme` option within loadBalancer.server which only has 'https' as an option. Perhaps we can have 'http3' as a value, to have the loadBalancer use HTTP3, to then allow webtransport to work?

I suppose this feature request is for http3 loadbalancer support for the http service, which should then let webtransport work over that.

**Comment by mmatur** [comment]
Hi @Jashepp 

Thanks for your suggestion.

We are interested in this issue, but we'd be interested in the traction it will receive. We are going to leave the status as kind/proposal to give the community time to let us know if they would like this idea.

We will reevaluate as people respond.

Conversation is time-boxed to 6 months.

**Comment by emilevauge** [comment]
@caniko As @mmatur explained earlier, we have a [community driven process](https://doc.traefik.io/traefik/contributing/submitting-issues/#feature-request) to prioritize proposals which requires users to interact with proposals to be promoted to the roadmap.
Additionaly, this proposal requires a very significant effort. Our resources are limited, we have to carefully assign our workforce.
If you plan to contribute on bringing HTTP/3 native proxying, we would love to review your work :)

**Comment by caniko** [comment]
All I am saying is that this community driven approach introduces the chicken and egg problem. You are debating adopting this feature, and in the meantime teams that need this stop using K8 for this side of their application.

I will be moving the HTTP3 part of my stack to VMs on GCE where there is full HTTP3 support, I don't have time for semantic debates. Wish you the best.

**Comment by emilevauge** [comment]
If you are not ready to contribute, you can't blame the team for being too slow for you. 
Good luck 🙂

**Comment by 0xF6** [comment]
Guys can you tell me the state?
It seems that quic-go has a [library for web transport](https://github.com/quic-go/webtransport-go).
It became interesting, I would have implemented and sent PR, but I'm too stupid for that

**Comment by nmengin** [comment]
Hello there,

Thank you for the feedback, we have set the status to enhancement.

However, this would not make it to our roadmap as we are focused elsewhere. If a community member would like to build it, let us know, and we will work with you to ensure you have all the information needed to merge it.

We prefer to work with our community members at the beginning of the design process to ensure that we are aligned and can move quickly with the review and merge process. Let us know here or create a PR before you start, and we will work with you there.

Don’t forget to check out the [contributor docs](https://github.com/traefik/contributors-guide/blob/master/pr_guidelines.md) and link the PR to this issue.

**Comment by HeroBrine1st** [comment]
TL;DR I added webtransport to traefik to successfully tunnel UDP and I can't understand what exactly community wants

I'm probably going to implement this. I currently have a private, non-production fork[^2] of traefik having a UDP tunnel using WebTransport coexisting with HTTP3 server by wrapping it into webtransport server and having the same handlers, except for new type of services called "datagram" services which instead of handling HTTP do Upgrade and then use webtransport for unreliable datagrams. I believe I've seen codebase enough that I can do that.

[^2]: I actually just fixed my integration test, it's not even evaluated for produciton yet, but I am familiar enough with codebase so you see this message

This passes my integration tests but I currently did not test it on regular HTTP/3 and even on other forms of traffic. I will eventually find a way to fix that, too. My end goal is ability to pass UDP through QUIC in a way not requiring any form of MITM, directly to traefik binary which does the handshake[^1]

[^1]: This excludes various approaches like using other reverse proxies that a number of people have already proposed to me

Before I even start asking maintainers, I want the following info:

> Hi, can we get support for WebTransport within the HTTP service, where we can provide a HTTP loadBalancer that connects to a HTTP 3 (Quic) server, which accepts WebTransport (or some other similar solution to get this working)?

As far as I know, QUIC cannot be used without TLS. This means that service accepting QUIC requests must do TLS handshake, i.e. in case of traefik with its routers it would require a form of MITM. I don't see this as correct implementation, so I think it's not what's being asked.

So, what exactly do the community want? My goal is to land a UDP router that works like TCP router with TLS, i.e. a UDP router that does TLS termination. This was the first thing that came to my mind when I saw "UDP routers" only to be disappointed that it's just a NAT/load balancer.

If community wants something that would allow creating arbitrary UDP tunnels without further patching (external programs are okay), I will be glad to implement it. Note that while I am a talented self-taught developer, I have literally 10 hours of experience with Go.

---

I also want to pinpoint a very disturbing problem in codebase that, I guess, needs to be fixed. This very type is what is actually passed to ServeHTTP methods:

https://github.com/traefik/traefik/blob/9893e89628425c91faa1fa7768eff8f1ccdc596f/pkg/middlewares/recovery/recovery.go#L64-L68

It hides http3 interfaces, disallowing services to use them. 

<details>
<summary>My approach is.. questionable</summary>

```go
func newRecoveryResponseWriter(rw http.ResponseWriter) recoveryResponseWriter {
	wrapper := &responseWriterWrapper{rw: rw}

	s, isST := rw.(http3.Settingser)
	h, isHM := rw.(http3.HTTPStreamer)
	cn, isCN := rw.(http.CloseNotifier)

	isH3 := isST && isHM

	switch {
	case isH3 && isCN:
		return &struct {
			*responseWriterWrapper
			http3.Settingser
			http3.HTTPStreamer
			http.CloseNotifier
		}{wrapper, s, h, cn}

	case isH3:
		return &struct {
			*responseWriterWrapper
			http3.Settingser
			http3.HTTPStreamer
		}{wrapper, s, h}

	case isCN:
		return &struct {
			*responseWriterWrapper
			http.CloseNotifier
		}{wrapper, cn}

	default:
		return wrapper
	}
}
```
That's of course not something I want to include in PR

</details>

Actually it is a nobrainer to remove CloseNotifier because it is deprecated and not used anymore in codebase (if I grepped correctly).

**Comment by ullebe1** [comment]
> As far as I know, QUIC cannot be used without TLS. This means that service accepting QUIC requests must do TLS handshake, i.e. in case of traefik with its routers it would require a form of MITM. I don't see this as correct implementation, so I think it's not what's being asked.

Isn't Traefik already effectively MITM'ing all connections, when used as a reverse proxy where it terminates all TLS for connections going to unencrypted upstreams? Anyway, I can't speak for others but I can share what I'm looking for as a "subscriber" to this issue:

I'd like to be able to do reverse proxying on WebTransport connections. Have all of them come in to Traefik to then be directed to the correct upstream container in a Kubernetes cluster. To me Traefik terminating the TLS connection and creating a new one when connecting to the upstream over Quic is an OK tradeoff, as Traefik is effectively already able to MITM all my normal traffic.

> My goal is to land a UDP router that works like TCP router with TLS, i.e. a UDP router that does TLS termination.

That definitely also sounds like a valuable addition! Just separate from what I'm looking to achieve.

**Comment by HeroBrine1st** [comment]
> Isn't Traefik already effectively MITM'ing all connections, when used as a reverse proxy where it terminates all TLS for connections going to unencrypted upstreams

Yeah, it just doesn't encrypt it again. I actually skimmed through RFCs and found there is probably no standard for reverse proxying, so if folks at Nginx or Caddy didn't come up with something entirely better we'd need to ignore TLS verification issues which I guess needs to be opt-in or someone will land a CVE eventually

It's also worrying that while I did patch webtransport server into pipeline, it uses the same ServeHTTP method - meaning one stream per backend connection. I hope it is not an issue due to the fact it is just a different connection ID so that it's the exact same network overhead and, most probably, instead of O(log n) (or whatever) on QUIC streams it is O(log n) on QUIC connections, while each connection has only one stream.

My current approach supports exactly that. It is a minimal change that's only invasive to recovery (recoveryResponseWrapper above) and accessLog (same issue in captureResponseWrapper, yeah I already started staging it).

A "complete" solution would require changing all load balancers of course (there is a chokepoint so it's one change) but it is essentialy "is it http3? Is it Upgrade: webtransport?" and then wtServer.upgrade(...) plus backend dial (and two goroutines to bridge streams together). The problem is only in wrappers which hide interfaces.

**Comment by HeroBrine1st** [comment]
This is one of chokepoints of all backend dials:

https://github.com/traefik/traefik/blob/9893e89628425c91faa1fa7768eff8f1ccdc596f/pkg/proxy/httputil/proxy.go#L54-L63

Which brings another approach by implementing httputil.ReverseProxy interface using quic-go. I checked, there's none currently available. They have reverse proxy in tests though. 

This still suffers from wrappers, but uses proper polymorphism for quic:// urls, so that all backend connections are upgraded to QUIC. This approach looks most clean

**Comment by Nelwhix** [comment]
Hi @nmengin, I would like to take a stab at implementing it

### Proposed design                                                                                                                                        
                                                                                                                                                                          
 **Phase 1 — HTTP/3 as a backend transport, HTTP semantics only (no WebTransport yet).**                                                                                
                                                                                                                                                                        
 - Add a `quic` (or `http3`) value to `Server.Scheme`, validated the same way                                                                                           
   `h2c` is today in each provider's scheme parsing.                                                                                                                    
 - Add an `http3.RoundTripper` (from `quic-go/http3`) as a fourth leg of                                                                                                
   `smartRoundTripper`, alongside `http`, `http2`, `h2c`, selected by scheme                                                                                            
   (QUIC does its own ALPN, so no protocol-sniffing is needed).                                                                                                         
 - Fix the two response-writer wrappers to preserve                                                                                                                     
   `http3.Settingser`/`http3.HTTPStreamer`.                                                                                                                                   
 - This alone unlocks plain HTTP/3-to-HTTP/3 backend proxying and is                                                                                                    
   independently useful and testable without touching WebTransport semantics                                                                                         
 - it could ship as its own PR.              

**Phase 2 — WebTransport session proxying.**                                                                                                                           
                                                                                                                                                                          
   - WebTransport's upgrade/stream/datagram model doesn't fit                                                                                                             
     `httputil.ReverseProxy`'s byte-stream assumptions, so treat it the way                                                                                               
     Traefik already treats CONNECT/websocket: a dedicated handler path that detects the                                                                                               
     WebTransport upgrade, opens a client-side WebTransport session to the                                                                                                
     backend (via `quic-go/webtransport-go`), and bridges streams/datagrams                                                                                               
     bidirectionally  rather than forcing it through `rewriteRequestBuilder`.                                                                                            
   - Traefik already TLS-terminates and re-encrypts to backends for                                                                                                       
     `https`/`h2c`; the same trust model applies here (as raised by @ullebe1) —                                                                                           
     this doesn't introduce a new MITM concern beyond what already exists for                                                                                             
     every HTTPS backend today.                                    

I have some questions though:
 - Is a new `Server.Scheme` value (`http3`/`quic`) the right way to go, or should                                                                                            
     this be a `ServersTransport`-level opt-in instead?                                                                                                                   
 - Should WebTransport support build on the existing HTTP service (new scheme + upgrade handling), or warrant its own router/service type, given it isn't a request/response protocol?
- Should it ship behind an experimental flag initially, given the WebTransport spec itself is still evolving?
